lib/swtpmtest.pm

```diff
@@ -0,0 +1,136 @@
+    },
+    swtpm_2 => {
+        xml_file    => {uefi => 'swtpm_uefi_2_0.xml', legacy => 'swtpm_legacy_2_0.xml'},
+        version     => '2.0',
+        ssh_script  => 'ssh_script_2',
+        result_file => '/tmp/swtpm_2',
+    },
+};
+
+my $sample_cfg = {
+    uefi   => {vm_name => 'SLES15SP3-UEFI-swtpm',   sample_file => 'swtpm_uefi.xml'},
+    legacy => {vm_name => 'SLES15SP3-LEGACY-swtpm', sample_file => 'swtpm_legacy.xml'},
+};
+
+# Function "start_swtpm_vm" starts a vm via libvirt "virsh" commands,
+# sample xml file and vm image file are pre-configured
+sub start_swtpm_vm {
+    my ($swtpm_ver, $swtpm_vm_type) = @_;
+    die "invalid swtpm type parameter $swtpm_ver"  unless ($guestvm_cfg->{$swtpm_ver});
+    die "invalid vm type parameter $swtpm_vm_type" unless ($sample_cfg->{$swtpm_vm_type});
+
+    # Copy the sample configuration file and modify it based on the test requirement
+    my $guest_swtpm_ver = $guestvm_cfg->{$swtpm_ver};
+    my $guest_mode      = $sample_cfg->{$swtpm_vm_type};
+    my $vm_name         = $guest_mode->{vm_name};
+    my $smaple_xml      = $guest_mode->{sample_file};
+    my $guest_xml       = $guest_swtpm_ver->{xml_file};
+    my $swtpm_type      = $guest_swtpm_ver->{version};
+    assert_script_run("cd $image_path");
+    assert_script_run("cp $smaple_xml $guest_xml->{$swtpm_vm_type}");
+    assert_script_run(
+"sed -i \"/<\\/devices>/i\\    <tpm model='tpm-tis'>\\n      <backend type='emulator' version='$swtpm_type'\\/>\\n    <\\/tpm>\" $guest_xml->{$swtpm_vm_type}"
+    );
+
+    # Define the guest vm and start it
+    assert_script_run("virsh define $guest_xml->{$swtpm_vm_type}");
+    assert_script_run("virsh start $vm_name");
+}
+
+# Function "stop_swtpm_vm" stops/destroys a running vm
+sub stop_swtpm_vm {
+    my $para = shift;
+    die "invalid vm type parameter $para" unless ($sample_cfg->{$para});
+
+    # For UEFI vm guest, we should add --nvram option
+    my $guest_mode = $sample_cfg->{$para};
+    my $vm_name    = $guest_mode->{vm_name};
+    assert_script_run("virsh destroy $vm_name");
+    if ($para eq "uefi") {
+        assert_script_run("virsh undefine $vm_name --nvram");
+    }
+    else {
+        assert_script_run("virsh undefine $vm_name");
+    }
+}
+
+# Function "swtpm_verify" gets ip address of the vm, and ssh access
+# into it, then check the tpm parameters
+sub swtpm_verify {
+    my $para = shift;
+    die "invalid swtpm parameter $para" unless ($guestvm_cfg->{$para});
+
+    # Check the vm guest is up via listening to the port 22
+    my $ssh_port_chk_script = 'ssh_port_chk_script';
+    assert_script_run("wget --quiet " . data_url("swtpm/$ssh_port_chk_script") . " -O $image_path/$ssh_port_chk_script");
+    assert_script_run("chmod 755 $image_path/$ssh_port_chk_script");
+    assert_script_run("sh $image_path/$ssh_port_chk_script");
+
+    # Login to the vm and run the commands to check tpm device
+    my $user   = 'root';
+    my $passwd = 'nots3cr3t';
```

Comment:
better not repeat the default password but reuse from variables

Reply:
Nice comment, will do

Reply:
done